jetty-core/jetty-deploy/src/main/java/org/eclipse/jetty/deploy/DeploymentScanner.java

```diff
@@ -0,0 +1,1305 @@
+                            appEnvironment = getDefaultEnvironmentName();
+                        app.setEnvironment(Environment.get(appEnvironment));
+
+                        // Create a new Attributes layer for the app deployment, which is the
+                        // combination of layered Environment Attributes with app Attributes overlaying them.
+                        Attributes envAttributes = environmentAttributesMap.get(appEnvironment);
+                        Attributes deployAttributes = envAttributes == null ? app.getAttributes() : new Attributes.Layer(envAttributes, app.getAttributes());
+
+                        // Ensure that Environment configuration XMLs are listed in deployAttributes
+                        List<Path> envXmlPaths = findEnvironmentXmlPaths(deployAttributes);
+                        envXmlPaths.sort(PathCollators.byName(true));
+                        PathsContextHandlerFactory.setEnvironmentXmlPaths(deployAttributes, envXmlPaths);
+
+                        // Create the Context Handler
+                        Path mainPath = app.getMainPath();
+                        if (mainPath == null)
+                            throw new IllegalStateException("Unable to create ContextHandler for app with no main path defined: " + app);
+                        ContextHandler contextHandler = contextHandlerFactory.newContextHandler(server, app.getEnvironment(), mainPath, app.getPaths().keySet(), deployAttributes);
+                        app.setContextHandler(contextHandler);
+
+                        // Introduce the ContextHandler to the Deployer
+                        startTracking(app);
+                        deployer.deploy(app.getContextHandler());
+                    }
+                }
+            }
+            catch (Throwable t)
+            {
+                LOG.warn("Failed to to perform action {} on {}", step.type(), app, t);
+                ExceptionUtil.ifExceptionThrowUnchecked(t);
+            }
+            finally
+            {
+                app.resetStates();
+            }
+        }
+
+        // Fully stop tracking apps that have been removed, but not re-added.
+        for (PathsApp removed : removedApps)
+        {
+            stopTracking(removed);
+        }
+    }
+
+    protected List<DeployAction> sortActions(List<DeployAction> actions)
+    {
+        Comparator<DeployAction> deployActionComparator = getActionComparator();
+        if (deployActionComparator != null)
+            actions.sort(deployActionComparator);
+        return actions;
+    }
+
+    private List<Path> findEnvironmentXmlPaths(Attributes deployAttributes)
+    {
+        List<Path> rawEnvXmlPaths = deployAttributes.getAttributeNameSet()
+            .stream()
+            .filter(k -> k.startsWith(ContextHandlerFactory.ENVIRONMENT_XML_ATTRIBUTE))
+            .map(k -> Path.of((String)deployAttributes.getAttribute(k)))
+            .toList();
```

Comment:
This should be treated as a Resource rather than a path, so that differences in OS can be handled with OS neutral configuration.

Reply:
Actually, do we even need this mechanism now that we have the `environments` directory?